Spec/Auth.swift

```diff
@@ -3,17 +3,113 @@ import Ably.Private
+                    if (options == nil) {
+                        options = AblyTests.commonAppSetup()
+                        options.clientId = currentClientId
+                        rest = ARTRest(options: options)
+                    }
+                }
+                private let json = "{" +
+                "    \"token\": \"xxxxxx.yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy\"," +
+                "    \"issued\": 1479087321934," +
+                "    \"expires\": 1479087363934," +
+                "    \"capability\": \"{\\\"test\\\":[\\\"publish\\\"]}\"," +
+                "    \"clientId\": \"myClientId\"" +
+                "}"
+
+                private func check(_ details: ARTTokenDetails) {
+                    expect(details.token).to(equal("xxxxxx.yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy"))
+                    expect(details.issued).to(equal(Date(timeIntervalSince1970: 1479087321.934)))
+                    expect(details.expires).to(equal(Date(timeIntervalSince1970: 1479087363.934)))
+                    expect(details.capability).to(equal("{\"test\":[\"publish\"]}"))
+                    expect(details.clientId).to(equal("myClientId"))
+                }
+            private let channelName = "test_JWT"
+            private let messageName = "message_JWT"
+                private let jwtTestsOptions = AblyTests.clientOptions()
+                private let authUrlTestsOptions: ARTClientOptions = {
+                    let options = AblyTests.clientOptions()
+                    options.authUrl = URL(string: echoServerAddress)!
+                    return options
+                }()
+
+                private var keys: [String: String]!
+
+                private func rsa8gTestsSetupDependencies() {
+                    if (keys == nil) {
+                        keys = getKeys()
+                    }
+                }
+                private let authCallbackTestsOptions = AblyTests.clientOptions()
+            private let jwtAndRestTestsOptions = AblyTests.clientOptions()
+                private var client: ARTRest!
+
+                private func jwtContentTypeTestsSetupDependencies() {
+                    if (client == nil) {
+                        let options = AblyTests.clientOptions()
+                        let keys = getKeys()
+                        options.authUrl = URL(string: echoServerAddress)!
+                        options.authParams = [URLQueryItem]()
+                        options.authParams?.append(URLQueryItem(name: "keyName", value: keys["keyName"]))
+                        options.authParams?.append(URLQueryItem(name: "keySecret", value: keys["keySecret"]))
+                        options.authParams?.append(URLQueryItem(name: "returnType", value: "jwt"))
+                        client = ARTRest(options: options)
+                    }
+                }
 
 class Auth : QuickSpec {
-    override func spec() {
+
+override class var defaultTestSuite : XCTestSuite {
```

Comment:
This `defaultTestSuite` override serves to make sure that all of the variables that were originally local variables at `spec` / `describe` / `context` level get initialized at the start of the test suite, to be consistent with the behaviour in Quick. We have some tests which will fail if this initialization order is not respected.

Why override `defaultTestSuite`? It's a bit of a hack, really. I just needed some method that'd be called at the start of the test suite, and I couldn't find any better override. I could have instead placed something _outside_ the individual test classes and used `XCTestObservation` to call some initialization method on each of the classes, but it would have meant that the migrator didn't output self-contained test files any longer. I think we can live with it, but if you disagree or would find it useful to add a comment to the code explaining what's going on, let me know.

We probably should fix this dependency on initialization order anyway, at which point we can get rid of this code.

Reply:
what about this https://developer.apple.com/documentation/xctest/xctestcase/1496262-setup ?

>The setUp() class method is called exactly once for a test case, before its first test method is called. Override this method to customize the initial state for all tests in the test case.

Reply:
The difference is that `setUp()` is called before the tests in _that file_ are run. To be consistent with the behaviour of Quick, I need to set up the global variables _for all test files_ before even the first test of the test suite runs. Does that make sense?

Reply:
It makes sense, but it's called four times from four different classes with different globals. In this case, I don't understand it.